File: src/GameOfLife.scala
object GameOfLife extends App {
  case class Cell(x: Int, y: Int)
  case class Rel(alive: Boolean, neighbors: Int)

  def cellTick(rel: Rel) = rel.neighbors match {
    case 2 => rel.alive
    case 3 => true
    case _ => false
  }

  def explode(coords: List[Cell]) = for {
    cell   <- coords
    deltax <- List(-1, 0, 1)
    deltay <- List(-1, 0, 1)
    center = (deltax == 0 && deltay == 0)
    neighbors = if (!center) 1 else 0
  } yield Map(Cell(cell.x + deltax, cell.y + deltay) -> Rel(center, neighbors))

  def cellMerge(a: Rel, b: Rel) =
    Rel((a.alive || b.alive), a.neighbors + b.neighbors)

  def worldTick(cells: List[Cell]) =
    explode(cells).
      foldLeft(Map[Cell, Rel]()) {
        case (sum, item) =>
          val key = item.keys.head
          val a = sum.getOrElse(key, Rel(false, 0))
          val b = item(key)
          sum ++ Map(key -> cellMerge(a, b))
      }.filter { case (cell, rel) => cellTick(rel) }.
      keys
}

// 4 Whitespace
// 3 Closing }
// 25 loc

File: src/game_of_life.erl
-module(game_of_life).
-export([world_tick/1]).

tick({A,2}) -> A;
tick({_,3}) -> true;
tick({_,_}) -> false.

explode(Coords) ->
    [[{{X+DeltaX, Y+DeltaY}, {Center, if Center -> 0; true -> 1 end}}
      || DeltaX <- [-1,0,1],
         DeltaY <- [-1,0,1],
         Center <- [DeltaX =:= 0 andalso DeltaY =:= 0]]
     || {X,Y} <- Coords].

cell_merge(_K, {A1,N1}, {A2,N2}) -> {A1 or A2, N1+N2}.

world_tick(Coords) ->
    Dict = lists:foldl(fun(D1,D2) -> dict:merge(fun cell_merge/3, D1, D2) end,
                       dict:new(),
                       [dict:from_list(L) || L <- explode(Coords)]),
    dict:fetch_keys(dict:filter(fun(_,V) -> tick(V) end, Dict)).

%% 4 whitespace
%% 17 loc
